codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/JsonSerializerGenerator.kt

```diff
@@ -82,14 +163,163 @@ class JsonSerializerGenerator(protocolConfig: ProtocolConfig) : StructuredDataSe
         val fnName = "serialize_document"
         return RuntimeType.forInlineFun(fnName, "operation_ser") {
             it.rustTemplate(
+                // TODO: Implement document parsing
                 """
                 pub fn $fnName(input: &#{Document}) -> Result<#{SdkBody}, #{Error}> {
-                    #{serde_json}::to_vec(&#{doc_json}::SerDoc(&input)).map(#{SdkBody}::from)
+                    unimplemented!();
+                }
+                """,
+                "Document" to RuntimeType.Document(runtimeConfig), *codegenScope
+            )
+        }
+    }
+
+    private fun RustWriter.serializeStructure(context: StructContext) {
+        val fnName = "serialize_structure_${context.shape.id.name.toSnakeCase()}"
+        val structureSymbol = symbolProvider.toSymbol(context.shape)
+        val structureSerializer = RuntimeType.forInlineFun(fnName, "json_ser") { writer ->
+            writer.rustBlockTemplate(
+                "pub fn $fnName(${context.objectName}: &mut #{JsonObjectWriter}, input: &#{Shape})",
+                "Shape" to structureSymbol,
+                *codegenScope,
+            ) {
+                if (context.shape.members().isEmpty()) {
+                    rust("let _ = input;") // Suppress an unused argument warning
+                }
+                for (member in context.shape.members()) {
+                    serializeMember(context.member(member))
+                }
+            }
+        }
+        rust("#T(${context.objectName.borrowMut()}, ${context.localName});", structureSerializer)
+    }
+
+    private fun RustWriter.serializeMember(context: MemberContext) {
+        val target = model.expectShape(context.shape.target)
+        handleOptional(context) { inner ->
+            val key = inner.keyExpression
+            val value = inner.valueExpression.borrow()
+            when (target) {
+                is StringShape -> when (target.hasTrait<EnumTrait>()) {
+                    true -> context.writeValue(this, "string", key, "$value.as_str()")
+                    false -> context.writeValue(this, "string", key, value)
+                }
+                is BooleanShape -> context.writeValue(this, "boolean", key, value)
+                is NumberShape -> {
+                    val numberType = when (symbolProvider.toSymbol(target).rustType()) {
+                        is RustType.Float -> "Float"
+                        is RustType.Integer -> "NegInt"
+                        else -> throw IllegalStateException("unreachable")
+                    }
+                    context.writeInner(this, "number", key) {
+                        rustInline("#T::$numberType(*${inner.valueExpression})", smithyTypes.member("Number"))
+                    }
+                }
+                is BlobShape -> context.writeInner(this, "string_unchecked", key) {
+                    rustInline("&#T($value)", RuntimeType.Base64Encode(runtimeConfig))
```

Comment:
you're welcome to use rustInline if you prefer, but we run `cargo fmt` on the output so it doesn't really matter if we write inline or not